File: backend/src/connectors/sqs.connector.ts
import {
  CreateQueueCommand,
  DeleteMessageCommand,
  GetQueueAttributesCommand,
  GetQueueUrlCommand,
  ReceiveMessageCommand,
  SendMessageCommand,
  SQSClient,
  type Message,
} from '@aws-sdk/client-sqs';
import { appConfig } from '../config';
import { logger } from '../logger';

let client: SQSClient | null = null;
const queueUrls = new Map<string, string>();

function getSqsClient(): SQSClient {
  if (!client) {
    client = new SQSClient({
      region: appConfig.sqs.region,
      endpoint: appConfig.sqs.endpoint,
      credentials: {
        accessKeyId: appConfig.sqs.accessKeyId,
        secretAccessKey: appConfig.sqs.secretAccessKey,
      },
    });
  }

  return client;
}

async function getQueueArn(queueUrl: string): Promise<string> {
  const response = await getSqsClient().send(new GetQueueAttributesCommand({
    QueueUrl: queueUrl,
    AttributeNames: ['QueueArn'],
  }));

  const arn = response.Attributes?.QueueArn;
  if (!arn) {
    throw new Error(`Queue ARN missing for ${queueUrl}`);
  }
  return arn;
}

export async function ensureQueueExists(queueName: string, isDlq = false): Promise<string> {
  const cached = queueUrls.get(queueName);
  if (cached) {
    return cached;
  }

  const sqs = getSqsClient();

  try {
    const response = await sqs.send(new GetQueueUrlCommand({ QueueName: queueName }));
    if (response.QueueUrl) {
      queueUrls.set(queueName, response.QueueUrl);
      return response.QueueUrl;
    }
  } catch {
    // Queue does not exist yet.
  }

  const attributes: Record<string, string> = {};
  if (!isDlq) {
    // The DLQ must exist before the main queue -- the redrive policy
    // references its ARN. A poison message stops cycling after
    // maxReceiveCount receives and lands there instead.
    const dlqUrl = await ensureQueueExists(`${queueName}-dlq`, true);
    attributes.VisibilityTimeout = String(appConfig.sqs.visibilityTimeoutSeconds);
    attributes.RedrivePolicy = JSON.stringify({
      deadLetterTargetArn: await getQueueArn(dlqUrl),
      maxReceiveCount: appConfig.sqs.maxReceiveCount,
    });
  }

  try {
    const createResponse = await sqs.send(new CreateQueueCommand({
      QueueName: queueName,
      Attributes: attributes,
    }));

    if (!createResponse.QueueUrl) {
      throw new Error(`Failed to create queue ${queueName}`);
    }

    queueUrls.set(queueName, createResponse.QueueUrl);
    logger.info(`Created SQS queue: ${queueName}`);
    return createResponse.QueueUrl;
  } catch (error) {
    const err = error as { name?: string };
    if (err.name === 'QueueNameExists' || err.name === 'QueueAlreadyExists') {
      const response = await sqs.send(new GetQueueUrlCommand({ QueueName: queueName }));
      if (!response.QueueUrl) {
        throw new Error(`Queue URL missing for ${queueName}`);
      }
      queueUrls.set(queueName, response.QueueUrl);
      return response.QueueUrl;
    }
    throw error;
  }
}

export async function ensureAllQueuesExist(): Promise<void> {
  const queueNames = Object.values(appConfig.sqs.queues);
  await Promise.all(queueNames.map((queueName) => ensureQueueExists(queueName)));
}

export async function sendQueueMessage(
  queueName: string,
  body: string
): Promise<string | undefined> {
  const url = await ensureQueueExists(queueName);
  const sqs = getSqsClient();

  const response = await sqs.send(
    new SendMessageCommand({
      QueueUrl: url,
      MessageBody: body,
    })
  );

  return response.MessageId;
}

export async function receiveQueueMessages(
  queueName: string,
  maxMessages = 1
): Promise<Message[]> {
  const url = await ensureQueueExists(queueName);
  const sqs = getSqsClient();

  const response = await sqs.send(
    new ReceiveMessageCommand({
      QueueUrl: url,
      MaxNumberOfMessages: maxMessages,
      WaitTimeSeconds: appConfig.sqs.waitTimeSeconds,
      VisibilityTimeout: appConfig.sqs.visibilityTimeoutSeconds,
    })
  );

  return response.Messages ?? [];
}

export async function deleteQueueMessage(
  queueName: string,
  receiptHandle: string
): Promise<void> {
  const url = await ensureQueueExists(queueName);
  const sqs = getSqsClient();

  await sqs.send(
    new DeleteMessageCommand({
      QueueUrl: url,
      ReceiptHandle: receiptHandle,
    })
  );
}

File: inference/src/connectors/sqs_connector.py
import json

import boto3
from botocore.exceptions import ClientError

from ..config import app_config
from ..logger import logger

_client = None
_queue_urls: dict[str, str] = {}

_NON_EXISTENT_CODES = (
    "AWS.SimpleQueueService.NonExistentQueue",
    "QueueDoesNotExist",
    "NonExistentQueue",
)


def _get_sqs_client():
    global _client
    if _client is None:
        sqs_config = app_config["sqs"]
        _client = boto3.client(
            "sqs",
            region_name=sqs_config["region"],
            endpoint_url=sqs_config["endpoint"],
            aws_access_key_id=sqs_config["access_key_id"],
            aws_secret_access_key=sqs_config["secret_access_key"],
        )
    return _client


def _get_queue_arn(queue_url: str) -> str:
    response = _get_sqs_client().get_queue_attributes(
        QueueUrl=queue_url, AttributeNames=["QueueArn"])
    return response["Attributes"]["QueueArn"]


def ensure_queue_exists(queue_name: str, is_dlq: bool = False) -> str:
    cached = _queue_urls.get(queue_name)
    if cached:
        return cached

    sqs = _get_sqs_client()

    try:
        response = sqs.get_queue_url(QueueName=queue_name)
        _queue_urls[queue_name] = response["QueueUrl"]
        return response["QueueUrl"]
    except ClientError as error:
        if error.response["Error"]["Code"] not in _NON_EXISTENT_CODES:
            raise

    attributes = {}
    if not is_dlq:
        # The DLQ must exist before the main queue -- the redrive policy
        # references its ARN. A poison message stops cycling after
        # max_receive_count receives and lands there instead.
        dlq_url = ensure_queue_exists(f"{queue_name}-dlq", is_dlq=True)
        attributes["VisibilityTimeout"] = str(
            app_config["sqs"]["visibility_timeout_seconds"])
        attributes["RedrivePolicy"] = json.dumps({
            "deadLetterTargetArn": _get_queue_arn(dlq_url),
            "maxReceiveCount": app_config["sqs"]["max_receive_count"],
        })

    try:
        response = sqs.create_queue(QueueName=queue_name, Attributes=attributes)
        logger.info(f"Created SQS queue: {queue_name}")
    except ClientError as error:
        if error.response["Error"]["Code"] not in ("QueueNameExists", "QueueAlreadyExists"):
            raise
        response = sqs.get_queue_url(QueueName=queue_name)

    _queue_urls[queue_name] = response["QueueUrl"]
    return response["QueueUrl"]


def send_queue_message(queue_name: str, body: str):
    url = ensure_queue_exists(queue_name)
    response = _get_sqs_client().send_message(QueueUrl=url, MessageBody=body)
    return response.get("MessageId")


def receive_queue_messages(queue_name: str, max_messages: int = 1) -> list:
    url = ensure_queue_exists(queue_name)
    sqs_config = app_config["sqs"]
    response = _get_sqs_client().receive_message(
        QueueUrl=url,
        MaxNumberOfMessages=max_messages,
        WaitTimeSeconds=sqs_config["wait_time_seconds"],
        VisibilityTimeout=sqs_config["visibility_timeout_seconds"],
    )
    return response.get("Messages", [])


def delete_queue_message(queue_name: str, receipt_handle: str) -> None:
    url = ensure_queue_exists(queue_name)
    _get_sqs_client().delete_message(QueueUrl=url, ReceiptHandle=receipt_handle)
